images/bootstrap/gcs_restapi.sh

```diff
@@ -0,0 +1,129 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Function to validate GOOGLE_APPLICATION_CREDENTIALS and and set access_token global variable
+get_access_token() {
+    if [ -z "${GOOGLE_APPLICATION_CREDENTIALS}" ]; then
+        echo "GOOGLE_APPLICATION_CREDENTIALS is not set. Please set it to the path of your service account key file."
+        exit 1
+    fi
+
+    local sa_email=$(jq -r '.client_email' "$GOOGLE_APPLICATION_CREDENTIALS")
+    local sa_key=$(jq -r '.private_key' "$GOOGLE_APPLICATION_CREDENTIALS")
+    local jwt_header=$(echo -n '{"alg":"RS256","typ":"JWT"}' | base64 -w 0 | tr '+/' '-_' | tr -d '=')
+    local jwt_claim=$(echo -n '{"iss":"'$sa_email'","scope":"https://www.googleapis.com/auth/cloud-platform","aud":"https://oauth2.googleapis.com/token","exp":'$(($(date +%s) + 3600))',"iat":'$(date +%s)'}' | base64 -w 0 | tr '+/' '-_' | tr -d '=')
+    local jwt_signature=$(echo -n "$jwt_header.$jwt_claim" | openssl dgst -binary -sha256 -sign <(echo "$sa_key") | base64 -w 0 | tr '+/' '-_' | tr -d '=')
+    local jwt="$jwt_header.$jwt_claim.$jwt_signature"
+
+    local response=$(curl -s -X POST https://oauth2.googleapis.com/token \
+         -H "Content-Type: application/x-www-form-urlencoded" \
+         -d "grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer&assertion=$jwt")
+
+    access_token=$(echo "$response" | jq -r '.access_token')
+
+    if [ -z "$access_token" ]; then
+        echo "Failed to obtain access token. Check your service account key file."
+        exit 1
+    fi
+}
+
+urlencode_path() {
+    local path="$1"
+    echo "$path" | sed 's/\//%2F/g'
+}
+
+# Function to upload a file to Google Cloud Storage
+upload_to_gcs() {
+    local source_file="$1"
+    local bucket_name="$2"
+    local destination_blob=$(urlencode_path "$3")
+    local content_type="application/octet-stream"
+
+    get_access_token
+
+    upload_response=$(curl -X POST \
+      --data-binary @"$source_file" \
+      -H "Authorization: Bearer $access_token" \
+      -H "Content-Type: $content_type" \
+      "https://storage.googleapis.com/upload/storage/v1/b/$bucket_name/o?uploadType=media&name=$destination_blob")
+
+    if echo "$upload_response" | jq -e '.name' > /dev/null; then
+       echo "File $source_file uploaded successfully as $destination_blob"
+       return 0
+    else
+       echo "Upload failed. Response:"
+       echo "$upload_response" | jq '.'
+       return 1
+    fi
+}
+
+# Function to check if a file exists in GCS
+stat_gcs_file() {
+    local bucket_name="$1"
+    local gcs_file_path=$(urlencode_path "$2")
+
+    get_access_token
+
+    stat_response=$(curl -s -X GET \
+      -H "Authorization: Bearer $access_token" \
+      "https://storage.googleapis.com/storage/v1/b/$bucket_name/o/$gcs_file_path")
```

Comment:
Do you think it makes sense to put `https://storage.googleapis.com/storage/v1/b/` into a variable? I see it reused a few times.

Reply:
I've two URLs
https://storage.googleapis.com/storage/v1/b/
https://storage.googleapis.com/upload/storage/v1/b

So I've just extracted as 
`BASE_URL="https://storage.googleapis.com"`